lib/fluent/plugin/buffer.rb

```diff
@@ -0,0 +1,351 @@
+
+      def metadata_list
+        synchronize do
+          @metadata_list.dup
+        end
+      end
+
+      def new_metadata(timekey: nil, tag: nil, variables: nil)
+        Metadata.new(timekey, tag, variables)
+      end
+
+      # TODO: metadata_list cleaner
+      def add_metadata(metadata)
+        synchronize do
+          if i = @metadata_list.index(metadata)
+            @metadata_list[i]
+          else
+            @metadata_list << metadata
+          end
+        end
+      end
+
+      def metadata(timekey: nil, tag: nil, variables: nil)
+        meta = new_meatadata(timekey: timekey, tag: tag, variables: variables)
+        add_metadata(meta)
+        meta
+      end
+
+      # metadata MUST have consistent object_id for each variation
+      # data MUST be Array of serialized events
+      def emit(metadata, data, force: false)
+        data_size = data.size
+        return if data_size < 1
+        raise BufferOverflowError unless storable?
+
+        stored = false
+
+        # the case whole data can be stored in staged chunk: almost all emits will success
+        chunk = synchronize { @stage[metadata] ||= generate_chunk(metadata) }
+        chunk.synchronize do
+          begin
+            chunk.append(data)
+            if !size_over?(chunk) || force
+              chunk.commit
+              stored = true
+            end
+          ensure
+            chunk.rollback
+          end
+        end
+        return if stored
+
+        # try step-by-step appending if data can't be stored into existing a chunk
+        emit_step_by_step(metadata, data)
+      end
+
+      def queued_records
+        synchronize { @queue.reduce(0){|r, chunk| r + chunk.records } }
+      end
+
+      def queued?(metadata=nil)
+        synchronize do
+          if metadata
+            n = @queued_num[metadata]
+            n && n.nonzero?
+          else
+            !@queue.empty?
+          end
+        end
+      end
+
+      def enqueue_chunk(metadata)
+        synchronize do
+          chunk = @stage.delete(metadata)
+          if chunk
+            chunk.synchronize do
+              if chunk.empty?
+                chunk.close
+              else
+                @queue << chunk
+                @queued_num[metadata] = @queued_num.fetch(metadata, 0) + 1
+                chunk.enqueued! if chunk.respond_to?(:enqueued!)
+              end
+            end
+          end
+          nil
+        end
+      end
+
+      def enqueue_all
+        synchronize do
+          if block_given?
+            @stage.keys.each do |metadata|
+              chunk = @stage[metadata]
+              v = yield metadata, chunk
+              enqueue_chunk(metadata) if v
+            end
+          else
+            @stage.keys.each do |metadata|
+              enqueue_chunk(metadata)
+            end
+          end
+        end
+      end
+
+      def dequeue_chunk
+        return nil if @queue.empty?
+        synchronize do
+          chunk = @queue.shift
```

Comment:
The first line check is early return. It's out of critical session (`synchronize`).
Other threads might dequeue chunks from this queue between first line and second line.